doc/design/requirements.md

```diff
@@ -0,0 +1,142 @@
+# Requirements and goals
+
+Includes the overlapping projects of TiUp, TiUp Cluster, and TiUp Playground.
+
+We should distinguish between goals and requirements for the long term vs an initial release.
+
+## Vision and high-level goals
+
+* Help expand the TiDB ecosystem by making it easy to share, install, and deploy components
+* Manage and deploy TiDB and TiKV clusters
+* Open source project, with a community of users and contributors
+* Help TiDB and TiKV developers run and test their code in a production-like environment
+* Make TiDB an easy to use database
+* Be the standard way for TiDB users to manage their clusters
+* Make it easy to try out TiDB, on a single machine and a cluster
+
+## Use cases
+
+* A new user with no experience with TiDB can try out TiDB with minimal friction.
+* A developer of TiDB or TiKV can setup a local environment to test, benchmark, and debug their code.
+* An oncall DBA or developer can quickly setup a cluster matching one where a bug was reported.
+* A user reporting a bug can provide clear steps to reproduce based on a TiUp cluster configuration
+* A TiDB DBA can manage a production cluster, including TODO
+
+
+## Requirements
+
+### Manage components
+
+* Install and update components
+* Uninstall/remove
+* Provide information about installed components and how to use them to the user
+* How must components be written? Restrictions on language, API? Sandboxing?
+* Components can be run directly from TiUp, components may run other components if they are dependencies (see versioning).
+  - E.g., running tidb can launch pd and tikv components
+
+#### Versioning
+
+* Components have an explicit version
+* Components may depend on other components and specify versions of those dependencies with which they are compatible
+* Components should use strict semver versioning
+* Component publishers will provide all versions
+* Version and semver compatibility is trusted, not verified
+* Users can specify which versions of each component to install
+* Component versions can be 'yanked', this means they are not listed and if the user tries to install them, they will be prompted for confirmation
+* Component versions cannot be deleted from servers
+* Component versions can be deleted locally (without removing the whole component)
+* Update
+  - patch revisions will be updated by default (i.e., the user must opt-in to keeping older versions)
+  - major and minor revisions will be duplicated by default (i.e., old versions are kept)
+  - when updating a component, by default, dependent components will be updated if necessary
```

Comment:
I don't think this introduces complexity, since we already need to keep track of dependencies for running the components or telling the user and we want the ability to update multiple components in any case too (update all), so supporting update of a dependency tree seems not too much extra. It is also fine for this to be low priority and to not support it for some time (I have not tried to prioritise requirements at this stage).

Reply:
Currently we have not tracked dependencies yet. TiUP doesn't know `playgrouond` dependent on `tidb/tikv/pd`. When you execute `tiup playground`, it's playground itself that download tidb/tikv/pd.

Reply:
How does playground know where to download from? Is there API in TiUp to get the location of the mirror?

I'll rephrase this to make it more implementation agnostic (i.e., it can be implemented in the component or TiUp).